src/plugins/intel_gna/src/transformations/gather_sinking_utils.cpp

```diff
@@ -0,0 +1,285 @@
+    ov::Rank::value_type max_input_rank = 0;
+    for (auto& input_node : node->input_values()) {
+        const ov::Rank output_rank = input_node.get_partial_shape().rank();
+        if (output_rank.is_dynamic())
+            return -1;
+        const ov::Rank::value_type output_rank_len = output_rank.get_length();
+        if (output_rank_len > max_input_rank)
+            max_input_rank = output_rank_len;
+    }
+    return max_input_rank;
+}
+
+NodePtr InsertUnsqueeze(Output<Node> node, size_t n_dims) {
+    std::vector<size_t> dims(n_dims);
+    std::iota(dims.begin(), dims.end(), 0);
+    auto unsqueeze_const = std::make_shared<Constant>(ov::element::i64, Shape{dims.size()}, dims);
+    auto unsqueeze = std::make_shared<Unsqueeze>(node, unsqueeze_const);
+    copy_runtime_info(node.get_node_shared_ptr(), {unsqueeze, unsqueeze_const});
+    return unsqueeze;
+}
+
+ov::Output<ov::Node> FixInputNodeRank(ov::Output<ov::Node> input_node, ov::Rank::value_type required_rank) {
+    const ov::Rank::value_type output_rank = input_node.get_partial_shape().rank().get_length();
+    if (output_rank >= required_rank)
+        return input_node;
+    return InsertUnsqueeze(input_node, required_rank - output_rank)->output(0);
+}
+
+ov::AxisVector ReverseGatherIndexes(const ov::AxisVector & indexes) {
+    ov::AxisVector positions(indexes.size());
+    for (size_t idx = 0; idx < indexes.size(); ++idx)
+        positions[indexes[idx]] = idx;
+
+    ov::AxisVector inverted(indexes.size());
+    for (size_t idx = 0; idx < indexes.size(); ++idx)
+        inverted[idx] = positions[idx];
+
+    return inverted;
+}
+
+std::shared_ptr<Constant> ReverseConstant(const std::shared_ptr<Constant>& node) {
+    if (!node)
+        return {};
+
+    const auto value = ReverseGatherIndexes(node->get_axis_vector_val());
+    const Shape& shape = node->get_output_shape(0);
+    const auto& element_type = node->get_element_type();
+
+    return std::make_shared<Constant>(element_type, shape, value);
+}
+
+}  // namespace
+
+namespace sink_backward {
+
+NodeVector InsertGatherBeforeNode(NodePtr main_node,
+                                  const std::shared_ptr<Constant>& indices_const,
+                                  const std::shared_ptr<Constant>& axes_const) {
+    if (HasDynamicRankInput(main_node))
+        return {};
+
+    NodeVector new_nodes;
+
+    const auto max_input_rank = GetMaxInputRank(main_node);
+    if (max_input_rank < 0)
+        return {};
+
+    for (size_t i = 0; i < main_node->get_input_size(); ++i) {
+        auto input_node = FixInputNodeRank(main_node->input_value(i), max_input_rank);
+
+        auto new_indices_const = indices_const->clone_with_new_inputs({});
+        auto new_axes_const = axes_const->clone_with_new_inputs({});
+        auto new_gather = std::make_shared<Gather>(input_node, new_indices_const, new_axes_const);
+
+        main_node->input(i).replace_source_output(new_gather->output(0));
```

Comment:
```suggestion
        node.replace_source_output(new_gather->output(0));
```

Reply:
It is not correct and fails building. input(i) returns Input<Node>. returns OutputVector = vector<Output<Node>>
So Output<Node> doesn't have replace_source_output

Reply:
Yeah then you can use inputs() it returns std::vector<Input<Node>> inputs()

Reply:
std::vector<Input> inputs() returns vector of Inputs but we need Outputs, that connected with that inputs. That fix is failed to build. In your way we need to find the Output of the input node that is connected with that Input. It is much simpler to go implemented way with indexes

Reply:
Output<Node> input_value(size_t input_index) const;

Reply:
We need Output for FixInputNodeRank while adding Unsqueeze.

Reply:
FixInputNodeRank returns Output<Node>